packages/ast/src/traverse.js

```diff
@@ -93,212 +93,22 @@ function walk(n: Node, cb: Cb, parentPath: ?NodePath<Node>) {
     return;
   }
 
-  switch (n.type) {
-    case "Program": {
-      const path = createPath(n, parentPath);
-      cb(n.type, path);
-
-      n.body.forEach(x => walk(x, cb, path));
-
-      break;
-    }
-
-    case "SectionMetadata":
-    case "FunctionNameMetadata":
-    case "ModuleNameMetadata":
-    case "LocalNameMetadata":
-    case "Data":
-    case "Memory":
-    case "Elem":
-    case "FuncImportDescr":
-    case "GlobalType":
-    case "NumberLiteral":
-    case "ValtypeLiteral":
-    case "FloatLiteral":
-    case "StringLiteral":
-    case "QuoteModule":
-    case "LongNumberLiteral":
-    case "BinaryModule":
-    case "LeadingComment":
-    case "BlockComment":
-    case "Identifier": {
-      cb(n.type, createPath(n, parentPath));
-      break;
-    }
-
-    case "ModuleExport": {
-      const path = createPath(n, parentPath);
-      cb(n.type, path);
-
-      walk(n.descr.id, cb, path);
-
-      break;
-    }
-
-    case "ModuleMetadata": {
-      const path = createPath(n, parentPath);
-      cb(n.type, path);
-      n.sections.forEach(x => walk(x, cb, path));
-
-      if (typeof n.functionNames !== "undefined") {
-        // $FlowIgnore
-        n.functionNames.forEach(x => walk(x, cb, path));
-      }
-
-      if (typeof n.localNames !== "undefined") {
-        // $FlowIgnore
-        n.localNames.forEach(x => walk(x, cb, path));
-      }
-      break;
-    }
-
-    case "Module": {
-      const path = createPath(n, parentPath);
-      cb(n.type, path);
-
-      if (typeof n.fields !== "undefined") {
-        n.fields.forEach(x => walk(x, cb, path));
-      }
-
-      if (typeof n.metadata !== "undefined") {
-        // $FlowIgnore
-        walk(n.metadata, cb, path);
-      }
-
-      break;
-    }
-
-    case "Start":
-    case "CallInstruction": {
-      const path = createPath(n, parentPath);
-      // $FlowIgnore
-      cb(n.type, path);
-
-      // $FlowIgnore
-      walk(n.index, cb, path);
-
-      break;
-    }
-
-    case "CallIndirectInstruction": {
-      const path = createPath(n, parentPath);
-      // $FlowIgnore
-      cb(n.type, path);
-
-      if (n.index != null) {
-        // $FlowIgnore
-        walk(n.index, cb, path);
-      }
-
-      break;
-    }
-
-    case "ModuleImport": {
-      cb(n.type, createPath(n, parentPath));
-
-      if (n.descr != null) {
-        // $FlowIgnore
-        walk(n.descr, cb, createPath(n, parentPath));
-      }
-
-      break;
-    }
-
-    case "Table":
-    case "Global": {
-      const path = createPath(n, parentPath);
-      cb(n.type, path);
-
-      if (n.name != null) {
-        walk(n.name, cb, path);
-      }
-
-      if (n.init != null) {
-        // $FlowIgnore
-        n.init.forEach(x => walk(x, cb, path));
+  const path = createPath(n, parentPath);
+  // $FlowIgnore
+  cb(n.type, path);
+
+  Object.keys(n).forEach((prop: string) => {
+    const value = n[prop];
+    if (!value) {
```

Comment:
Following my comment, I think that this check is not needed. Also it could lead to confusing like with node like `{hasFoo: false}`.

Reply:
@xtuc this check is needed (although to your point on `false` is not implemented correctly), because currently the AST has a number of [maybe types](https://flow.org/en/docs/types/primitives/#toc-maybe-types). I do wonder whether these should be changed to [optional types](https://flow.org/en/docs/types/primitives/#toc-optional-object-properties) instead for consistency?

Reply:
Yes it's inconsistent because sometimes I wanted to always have a key even null. I don't mind if we switch to optional types instead.

But I think both can be covered with `value != null`.